# --- Day 1: Historian Hysteria ---

Pair up the smallest number in the left list with the smallest number in the right list, then the second-smallest left number with the second-smallest right number, and so on.

Within each pair, figure out how far apart the two numbers are; you'll need to add up all of those distances.

## SAMPLE

We're gonna run the sample first. We will open the file and loop through the lines while printing each line.

In [21]:
sampleFile = "sample.txt"

In [22]:
with open(sampleFile) as file:
    for line in file:
        print(line.rstrip())

3   4
4   3
2   5
1   3
3   9
3   3


Let's now replace the spaces in each line with a comma and then split on the comma.

In [26]:
sampleFile = "sample.txt"
sampleCol1 = []
sampleCol2 = []

with open(sampleFile) as file:
    for line in file:
        linebits = line.replace("   ", ",").strip().split(",")
        sampleCol1.append(int(linebits[0]))
        sampleCol2.append(int(linebits[1]))

print (sampleCol1)
print (sampleCol2)

[3, 4, 2, 1, 3, 3]
[4, 3, 5, 3, 9, 3]


Sort each list

In [27]:
sampleCol1.sort()
sampleCol2.sort()

print(sampleCol1)
print(sampleCol2)

[1, 2, 3, 3, 3, 4]
[3, 3, 3, 4, 5, 9]


Loop through the lists using an index and add up the deltas.

In [29]:
totalDelta = 0
for idx, x in enumerate(sampleCol1):
    totalDelta += sampleCol2[idx] - x

print(totalDelta)

11


11 is the sample value so we know this code works. Let's run the same logic for the first input.

## PART 1

In [33]:
inputOne = "input-1.txt"
col1 = []
col2 = []
totalDelta = 0
simpleDelta = 0

with open(inputOne) as file:
    for line in file:
        linebits = line.replace("   ", ",").strip().split(",")
        col1.append(int(linebits[0]))
        col2.append(int(linebits[1]))

col1.sort()
col2.sort()

for idx, x in enumerate(col1):
    high = max(x, col2[idx])
    low = min(x, col2[idx])
    totalDelta += high - low
    simpleDelta += col2[idx] - x

print(totalDelta, simpleDelta)

1603498 970304


1603498 was the right answer for my input so moving onto part 2 of day 1. I suspected there would be some trick to the math because the puzzle talked about the difference between the numbers so I added the high/low bit before confirming this. Afterwards I went back and added the sample delta just to verify my concerns.

## PART 2

This time, you'll need to figure out exactly how often each number from the left list appears in the right list. Calculate a total similarity score by adding up each number in the left list after multiplying it by the number of times that number appears in the right list.

Input is the same

We'll proceed similarly as before by reading in the file and sorting the two lists. From there we should probably do a single loop through each list.

In [51]:
inputOne = "sample.txt"
col1 = []
col2 = []
col2dict = {}
similarityScore = 0

with open(inputOne) as file:
    for line in file:
        linebits = line.replace("   ", ",").strip().split(",")
        col1.append(int(linebits[0]))
        col2.append(int(linebits[1]))

col1.sort()
col2.sort()

print(col1)
print(col2)

[1, 2, 3, 3, 3, 4]
[3, 3, 3, 4, 5, 9]


We'll start by looping through the second column and creating a dictionary for each value and the value being the number of times that value appears. Then we'll loop through the first list and do the math.

In [52]:
col2dict.setdefault(0)
for idx, x in enumerate(col2):
    value = col2dict.get(x)
    col2dict[x] = 1 + (value if value != None else 0)

for idx, x in enumerate(col1):
    value = col2dict.get(x)
    similarityScore += x * (value if value != None else 0)

print(similarityScore)

31


#### Part 2 with real input file

In [54]:
inputOne = "input-1.txt"
col1 = []
col2 = []
col2dict = {}
similarityScore = 0

with open(inputOne) as file:
    for line in file:
        linebits = line.replace("   ", ",").strip().split(",")
        col1.append(int(linebits[0]))
        col2.append(int(linebits[1]))

col1.sort()
col2.sort()

for idx, x in enumerate(col2):
    value = col2dict.get(x)
    col2dict[x] = 1 + (value if value != None else 0)

for idx, x in enumerate(col1):
    value = col2dict.get(x)
    similarityScore += x * (value if value != None else 0)

print(similarityScore)

25574739


25574739 is the correct answer!